### ISSUE TITLE
LAM boundary rollout incorrectly implemented

### ISSUE THREAD

**Issue by dietervdb-meteo** [ecmwf/anemoi-core#212]
### What happened?

During rollout with a lam model, after a first 6h forecast step, the data outside of the cutout region needs to be replaced with data from the training dataset (rather than remain a forecast, as inside the cutout). This is currently incorrectly implemented in anemoi-training due to a misinterpretation of the workings of `torch.Tensor.masked_scatter`.  

### What are the steps to reproduce the bug?

I will present here a toy example that illustrates the issue. I have verified myself that this toy example correctly reproduces what is going wrong in actual anemoi training runs by saving some of the relevant tensors to disk, and checking that `boundary_update_training` as defined below reproduces the boundary replacement as performed by anemoi-training.

Let me first point to the relevant parts of the actual anemoi code. Boundary rollout as performed by anemoi training can be found [here](https://github.com/ecmwf/anemoi-core/blob/833e0e7ea905dcd8f38800ef3f49747282ddc81f/training/src/anemoi/training/utils/masks.py#L60). Boundary rollout as performed in anemoi-inference can be found [here](https://github.com/ecmwf/anemoi-inference/blob/c645fb219ed40ea60389affafe0f4f6c3b799ee9/src/anemoi/inference/runner.py#L759).

The difference between the two methods, and why the method used in training is erroneous can be seen by running the following simple python script:
```python
import numpy as np
import torch

def boundary_update_inference(input, fill_value, spatial_mask):
    updated_input = input.copy()
    updated_input[spatial_mask] = fill_value[spatial_mask]
    return updated_input

def boundary_update_training(input, fill_value, spatial_mask):
    updated_input = input.copy()
    x = torch.tensor(updated_input)
    fill_value = torch.tensor(fill_value)
    spatial_mask = torch.tensor(spatial_mask)
    x = x.masked_scatter(spatial_mask, fill_value)
    return x.numpy()

test = np.array([-1 for i in range(10)])
mask = np.array([ i > 3 for i in range(10)])
fill = np.array([ i for i in range(10)])

print("consider the following toy example (i <= 3 is 'cutout', i > 3 is 'boundary')")
print(f"test: {test}")
print(f"fill: {fill}")
print(f"boundary mask: {mask}")
print(f"replacement of test with fill on 'boundary' as in inference [correct]: {boundary_update_inference(toy_test,toy_fill,toy_mask)}")
print(f"replacement of test with fill on 'boundary' as in training [incorrect]: {boundary_update_training(toy_test,toy_fill,toy_mask)}")
```
The output of this script is
```
consider the following toy example (i <= 3 is 'cutout', i > 3 is 'boundary')
test: [-1 -1 -1 -1 -1 -1 -1 -1 -1 -1]
fill: [0 1 2 3 4 5 6 7 8 9]
boundary mask: [False False False False  True  True  True  True  True  True]
replacement of test with fill on 'boundary' as in inference [correct]: [-1 -1 -1 -1  4  5  6  7  8  9]
replacement of test with fill on 'boundary' as in training [incorrect]: [-1 -1 -1 -1  0  1  2  3  4  5]
```
One sees that `boundary_update_training` actually replaces the value of test at the first boundary gridpoint (index 4 in this example) with the value of fill at the first gridpoint (index 0 in this example) which is actually inside the cutout region! On the contrary  `boundary_update_inference` replaces the value of test at the first boundary gridpoint with the value of fill at the same gridpoint, which is what boundary rollout is supposed to do.

As mentioned above I have checked that the exact same problem can be observed when saving the tensors produced during training to disk and analysing them.

### Version

latest version

### Platform (OS and architecture)

irrelevant

### Relevant log output

```shell
You wont see this by looking at training logs.
```

### Organisation

RMI

**Comment by I-Dhar** [comment]
Hi dietervdb, I've been training a Australia wide LAM using AnemoI with 3 hour timestep which appeared to be working (see below) but when I run inference the results seem quite bad. For T2m, a 3 hour forecast has RMSE of about 2K (I was expecting RMSE of about 0.5K). Could this bug explain the poor results?

![Image](https://github.com/user-attachments/assets/2ba53a83-7772-4acd-b4d9-159b6d66fd8c)

Thanks

**Comment by dietervdb-meteo** [comment]
Hi Imtiaz, thanks for sharing. Yes, we observed the same, when looking at `wmse` during training things look good/reasonable (both `train_` and `val_wmse` going down) but in inference things look very bad, getting worse with increasing lead time. I'm 99% sure this bug is the reason [the model was replacing values on the boundary forecast with values from the interior reanalysis rather than boundary reananalysis]. PR #213  should fix this. We did some thorough sanity checking of the fix already, I'm currently training a model on rollout with it. Will then check if inference gives expected results. I expect the fix to be merged soon, but feel free to experiment with it already.

**Comment by dietervdb-meteo** [comment]
It has now been explicitly checked that by using the fixed code rollout training for lam models no longer deteriorates inference results at later lead times but rather improves them, as it should.
